Cipher fix, cipher test, and a note...

davidiw opened the pull request:
I can break this up into individual pull requests if there are any issues with any of the commits.

bford commented on README.md:
```diff
@@ -67,6 +67,16 @@ of NIST-standardized elliptic curves and symmetric cryptosystems.
+Issues
+------
+
+- Traditionally, ECDH uses only the x point for the shared secret. In this
```
This text is very unclear and out-of-the-blue; it needs some context, and in particular it's not obvious what exactly "uses" means here.  Can you expand the term "ECDH", and point to the most-relevant document specifying ECDH's standard behavior.  And you need to clarify here that "uses" means "forms the Diffie-Hellman shared secret from..."

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
 // Benchmark a block cipher operating in counter mode.
+/*
+XXX Broken
 func BlockCipherBench(b *testing.B, keylen int,
```
It looks like the Travis build is failing because this function is referenced but its definition is missing?

davidiw replied:
Interesting, I'll take a second look, but from what I can recall, this code wasn't compiling.

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+		return -1
+	}
+
+	mask1 := byte(1)
```
Argh, you can do better. :)  Don't unroll all this stuff 8 times here and in the for loop; just use nested loops with a bit of trivial arithmetic and the code size will be cut in half.

davidiw replied:
I did it because I'm willing to bet $1 that this is faster albeit less elegant.

bford replied:
Oh geesh, you're working about unrolling and micro-optimizing a function that's only used in a *test suite*???

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+
+	count := 0
+	for i := 0; i < len(a); i++ {
+		if (a[i] & mask1) != (b[i] & mask1) {
```
try if ((a[i] ^ b[i]) & (1 << j)) != 0 where j is the bit number from the inner loop

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+}
+
+// Tests a Cipher can encrypt and decrypt
+func BCHelloWorldHelper(t *testing.T,
```
Where does this name come from?  What is "BC"?

davidiw replied:
BlockCipher

On Tue, Feb 24, 2015 at 10:22 AM, Bryan Ford <notifications@github.com>
wrote:

> In test/cipher.go
> <https://github.com/DeDiS/crypto/pull/38#discussion_r25259682>:
>
> > +		if (a[i] & mask6) != (b[i] & mask6) {
> > +			count += 1
> > +		}
> > +		if (a[i] & mask7) != (b[i] & mask7) {
> > +			count += 1
> > +		}
> > +		if (a[i] & mask8) != (b[i] & mask8) {
> > +			count += 1
> > +		}
> > +	}
> > +
> > +	return float64(count) / float64(len(a)*8)
> > +}
> > +
> > +// Tests a Cipher can encrypt and decrypt
> > +func BCHelloWorldHelper(t *testing.T,
>
> Where does this name come from? What is "BC"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DeDiS/crypto/pull/38/files#r25259682>.
>

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+// 2) Encryption / decryption with different key don't work
+// 3) Changing a bit in the ciphertext or mac results in failed mac check
+// 4) Different keys produce sufficiently random output
+func AuthenticateAndEncrypt(t *testing.T,
```
"TestAuthenticatedEncryption"?

davidiw replied:
I avoided starting anything with a Test name since it seems to be special
to Go's unit test. I suspect the signature would likely make sure that Go
didn't get confused, but someone reading the code might.

On Tue, Feb 24, 2015 at 10:24 AM, Bryan Ford <notifications@github.com>
wrote:

> In test/cipher.go
> <https://github.com/DeDiS/crypto/pull/38#discussion_r25259871>:
>
> > +
> > +			res := BitDiff(ncrypts[i], ncrypts[j])
> > +			if res < bitdiff {
> > +				t.Log("Encryptions not sufficiently different:", res)
> > +				t.FailNow()
> > +			}
> > +		}
> > +	}
> > +}
> > +
> > +// Tests a Cipher:
> > +// 1) Encryption / decryption work
> > +// 2) Encryption / decryption with different key don't work
> > +// 3) Changing a bit in the ciphertext or mac results in failed mac check
> > +// 4) Different keys produce sufficiently random output
> > +func AuthenticateAndEncrypt(t *testing.T,
>
> "TestAuthenticatedEncryption"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DeDiS/crypto/pull/38/files#r25259871>.
>

bford replied:
Indeed, I think Go will avoid getting confused due to the function signature, and people can probably take a hint from the function signature as well. :)  (But I've seen the practice of attaching "...Helper" to the name and stuff like that to clarify it further.)

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+
+	// Encrypt / decrypt / mac test
+	for i := range nkeys {
+		nkeys[i] = make([]byte, keysize)
```
Looks like there's a bunch of key-setup code that's getting duplicated in each test; factor it out.

davidiw replied:
I had thought about it when writing. I'm almost in favor of completely removing the Hello, World test, since it seems quite redundant anyway. But factoring out the code in this specific scenario seems like it may result in more work then we really want to invest. I.e., should we go as far to construct a struct that contains an array of keys, encrypted messages, macs, etc?

bford replied:
Global variables are perfectly fine in a _test.go package.  Those don't get compiled into the non-test version of package so no need to worry about code bloat or startup performance due to them.

davidiw replied:
I understand the above comment, but I don't see the application in this context. The code sample has some redundancy with a far simpler test case in another function. We initialize variables each time, so that we can make the function generalizable to different ciphers, cleartext, and number of keys tested in parallel. Using global variables would not reduce the amount of initialization while keeping this generality.

bford replied:
Sorry, maybe I'm being too nit-picky. :)  No big deal.

davidiw replied:
After more careful examination, I realized I may have been misunderstanding you. You want the key generation to be done once in a separate init style function and store the results to global variables. I can easily implement that, but then we'd have a separate helper function that performs basically the following:

```
var keys [][]byte
func TestCipherInit(key_count int, key_size int) {
  keys = make([][]byte, key_count)
  for i := 0; i < key_count; i++ {
    keys[i] = make([]byte, key_size)
    rand.Read(keys[i])
  }
}
```

We'd still need to zero out the ncrypts and nmacs in each test case, since values are Xored with existing content and in not assigned. Perhaps I'm missing something, but I think this type of change would make the test code harder to read and it doesn't seem to be very critical to performance.

bford replied:
Yes, that's something like what I had in mind - but use your judgment; if it doesn't feel like it would improve conciseness/clarity then don't bother with it.

davidiw replied:
I feel mixed about it, because I had the same reaction as I was writing it and similar thoughts during our discussion. I tend to prefer test code that would guide users on how to use the code or that matches my intuition on good coding practice. So if we wanted to really invest the time into making this better, we ought to pass in the keys and other init variables as parameters into the function, but then I feel like I'm starting to do the same level of optimization as I did with the bit comparison :).

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+	for i := range ncrypts {
+		copy(ncrypts[i], deltacopy)
+
+		deltacopy[0] ^= 255
```
A few comments please about what's going on from one code block to the next?

davidiw replied:
I'm not sure I follow the exact location of this comment. There are a handful of comments before each block and if the code isn't self-documenting, then it likely ought to be improved. Of course, it could be a third choice which is that regardless of how elegant the code is, it still needs more documenting :).

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+		}
+		deltacopy[0] = ncrypts[i][0]
+
+		deltacopy[len(deltacopy)/2-1] ^= 255
```
Looks like each of these blocks of code is mostly duplicate but testing a slightly different case; factor out the common code into a separate function.  Doesn't need to be a separate top-level function if an internal higher-order function def would be useful, i.e.,

```
deltatest := func (...) { ...}
```

very handy because you can also refer to internal variables in the outer test function from the inner higher-order function.

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+	AuthenticateAndEncrypt(t, newCipher, n, bitdiff, mb)
+}
 
+func BlockCipherTest(t *testing.T,
```
Why "TestBlockCipher"?  These aren't block ciphers we're testing; we're testing the "abstract.Cipher" interface, which I've been calling a "Message Cipher" interface although that may not be the best name - but in any case the name "block cipher" is certainly not appropriate here.  For now just say TestCipher probably.

davidiw commented:
I've addressed most of your comments in the code and when not, I've left comments above. Let me know what you think. If you're happy, let me know, I'll squash and rebase my commits.

bford commented on README.md:
```diff
@@ -67,6 +67,17 @@ of NIST-standardized elliptic curves and symmetric cryptosystems.
+- Traditionally, ECDH (Elliptic curve Diffie-Hellman) derives the shared secret
+from the x point only. In this framework, you can either manually retrieve
+the value or take the combined (x, y) value. The latter is the recommended
+process as it can be done without loss of security and done more cleanly,
```
Avoid passive voice in technical writing!  (ALWAYS!!! :) )  "is the recommended process" - recommended by whom, and in what context?  And why can the latter be done more cleanly?  I would say:

```...or use the MarshalBinary method to take the combined (x, y) value as the shared secret.  We recommend the latter process for new software/protocols using this framework as it is cleaner and generalizes across different types of groups (e.g., both integer and elliptic-curve groups), ...```

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,230 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+	for i := 0; i < len(a); i++ {
+		for j := 0; j < 8; j++ {
+			if ((a[i] ^ b[i]) & (1 << uint(j))) != 0 {
+				count += 1
```
count++.

Or better yet, replace the whole if statement with:

```count += ((a[i] ^ b[i]) >> uint(j)) & 1```

davidiw replied:
Sorry, I've used Python too much recently :)

bford commented on test/cipher.go:
```diff
@@ -21,25 +24,230 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
+		}
+	}
+
+	for i := range ncrypts {
```
Some comments in here please?

davidiw commented:
I've squashed the previous commits and these new (rebased / squashed) commits should address all concerns.

bford commented:
Sounds good, thanks David!
